File: app/src/main/java/com/checkpoint/qr/detector/direction/DirectionDetectorHandler.java
package com.checkpoint.qr.detector.direction;

import android.graphics.Point;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.util.Pair;
import com.checkpoint.qr.detector.events.OnFinishDetectionEvent;
import com.checkpoint.qr.detector.events.OnPostProccessingNotDetectedEvent;
import com.checkpoint.qr.detector.pojos.DetectionResult;
import com.checkpoint.qr.detector.rest.QRDetectionEndPoint;

import java.util.List;

public class DirectionDetectorHandler  extends Handler {

    private static final String LOGTAG = DirectionDetectorHandler.class.getSimpleName();


    public DirectionDetectorHandler(final Looper looper) {
        super(looper);

    }


    @Override
    public void handleMessage(final android.os.Message msg) {
        final DirectionDetectorHandler.Message messageType = DirectionDetectorHandler.Message.fromOrdinal(msg.what);

        switch (messageType) {
            case CALCULATE_DIRECTION:
                Log.e("LAUNCH POSITION DETECTOR","---->");

                List<DirectionDetection> directionDetectionList = (List<DirectionDetection>) msg.obj;
                Log.e("----->",""+directionDetectionList.size());

                if(directionDetectionList.size()>0){
                  int size = directionDetectionList.size()-1;
                  Point center = directionDetectionList.get(0).center;
                  Point firstLocation = directionDetectionList.get(0).position;
                  Point secondLocation = directionDetectionList.get(size).position;
                  Log.e("center",""+center.toString());
                  Log.e("firstLocation",""+firstLocation.toString());
                  Log.e("secondLocation",""+secondLocation.toString());

                  if(firstLocation.x < secondLocation.x ){

                      new OnFinishDetectionEvent("DOWN TO UP",directionDetectionList.get(0).translate).broadcastEvent();

                  }
                  else if(firstLocation.x > secondLocation.x ){

                      new OnFinishDetectionEvent("UP to DOWN",directionDetectionList.get(0).translate).broadcastEvent();

                  }else{

                      if(secondLocation.x > center.x){

                          new OnFinishDetectionEvent("DOWN TO UP",directionDetectionList.get(0).translate).broadcastEvent();

                      }else{

                          new OnFinishDetectionEvent("UP to DOWN",directionDetectionList.get(0).translate).broadcastEvent();

                      }
                  }
                }


                break;
            default:
                Log.e(LOGTAG, "Unknown message received on DirectionDetectorHandlerThread");
        }
    }

    /**
     * Enum defining the messages that the BitmapHandlerThread can process.
     */
    public enum Message {
        CALCULATE_DIRECTION;
        private static final DirectionDetectorHandler.Message[] VALUES = values();
        private static DirectionDetectorHandler.Message fromOrdinal(final int ordinal) {
            return VALUES[ordinal];
        }
    }

}

File: app/src/main/java/com/checkpoint/qrdetector/handlers/DirectionDetectorHandler.kt
package com.checkpoint.qrdetector.handlers

import android.graphics.Point
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.checkpoint.qrdetector.events.OnFinishDetectionEvent
import com.checkpoint.qrdetector.model.DirectionDetection

/*
* Handler process to analyze a list of objects in order to determinate direction move
* of detected QR code, the remote parameter is a List of DirectionDetection object class
*
* */


class DirectionDetectorHandler(looper: Looper) : Handler(looper) {

     override fun handleMessage(msg: android.os.Message) {
         when (Message.fromOrdinal(msg.what)) {
            Message.CALCULATE_DIRECTION -> {
                val directionDetectionList = msg.obj as List<DirectionDetection>
                if (directionDetectionList.isNotEmpty()) {
                    val size = directionDetectionList.size - 1
                    val center: Point? = directionDetectionList[0].center
                    val firstLocation: Point? = directionDetectionList[0].position
                    val secondLocation: Point? = directionDetectionList[size].position
                    if (firstLocation!!.x < secondLocation!!.x) {
                        OnFinishDetectionEvent(
                            "DOWN TO UP",
                            directionDetectionList[0].translate,
                            directionDetectionList[0].image!!
                        ).broadcastEvent()
                    } else if (firstLocation.x > secondLocation.x) {
                        OnFinishDetectionEvent(
                            "UP to DOWN",
                            directionDetectionList[0].translate,
                            directionDetectionList[0].image!!
                        ).broadcastEvent()
                    } else {
                        if (secondLocation.x > center!!.x) {
                            OnFinishDetectionEvent(
                                "DOWN TO UP",
                                directionDetectionList[0].translate,
                                directionDetectionList[0].image!!
                            ).broadcastEvent()
                        } else {
                            OnFinishDetectionEvent(
                                "UP to DOWN",
                                directionDetectionList[0].translate,
                                directionDetectionList[0].image!!
                            ).broadcastEvent()
                        }
                    }
                }
            }
        }
    }
    enum class Message {
        CALCULATE_DIRECTION;

        companion object {
            private val VALUES = values()
            fun fromOrdinal(ordinal: Int): Message {
                return VALUES[ordinal]
            }
        }
    }

}